Source/WebCore/platform/graphics/gbm/MemoryMappedGPUBuffer.cpp

```diff
@@ -0,0 +1,444 @@
+        }
+
+        return useFormat;
+    };
+
+    auto bufferFormat = negotiateBufferFormat();
+    if (!bufferFormat.has_value()) {
+        LOG_ERROR("MemoryMappedGPUBuffer::create(), failed to negotiate buffer format");
+        return nullptr;
+    }
+
+    auto gbmBuffer = makeUnique<MemoryMappedGPUBuffer>(size);
+    if (!gbmBuffer->allocate(device, bufferFormat.value())) {
+        LOG_ERROR("MemoryMappedGPUBuffer::create(), failed to create GBM buffer of size %dx%d: %s", size.width(), size.height(), safeStrerror(errno).data());
+        return nullptr;
+    }
+
+    if (!gbmBuffer->createDMABufFromGBMBufferObject()) {
+        LOG_ERROR("MemoryMappedGPUBuffer::create(), failed to create dma-buf from GBM buffer object");
+        return nullptr;
+    }
+
+    return gbmBuffer;
+}
+
+bool MemoryMappedGPUBuffer::allocate(struct gbm_device* device, const GLDisplay::DMABufFormat& bufferFormat)
+{
+    m_modifier = DRM_FORMAT_MOD_INVALID;
+    if (!bufferFormat.modifiers.isEmpty())
+        m_bo = gbm_bo_create_with_modifiers2(device, m_size.width(), m_size.height(), bufferFormat.fourcc, bufferFormat.modifiers.data(), bufferFormat.modifiers.size(), GBM_BO_USE_RENDERING);
+
+    if (m_bo) {
+        m_modifier = gbm_bo_get_modifier(m_bo);
+
+        if constexpr (kDumpBufferAllocation)
+            LOG(GPUBuffers, " -> buffer allocated with modifiers (gbm_bo reports modifier 0x%02lx)", m_modifier);
+    } else {
+        m_bo = gbm_bo_create(device, m_size.width(), m_size.height(), bufferFormat.fourcc, GBM_BO_USE_LINEAR);
+        m_modifier = DRM_FORMAT_MOD_INVALID;
+
+        if constexpr (kDumpBufferAllocation)
+            LOG(GPUBuffers, " -> buffer allocated without modifiers (gbm_bo: 0x%2lx, m_modifier 0x%2lx)", gbm_bo_get_modifier(m_bo), m_modifier);
+    }
+
+    if (!m_bo) {
+        if constexpr (kDumpBufferAllocation)
+            LOG(GPUBuffers, " -> buffer allocation failed!");
+        return false;
+    }
+
+    m_format = gbm_bo_get_format(m_bo);
+    int planeCount = gbm_bo_get_plane_count(m_bo);
+    if (planeCount <= 0) {
+        if constexpr (kDumpBufferAllocation)
+            LOG(GPUBuffers, " -> invalid number of planes!");
+        return false;
+    }
+
+    m_planeCount = static_cast<unsigned>(planeCount);
+    return true;
+}
+
+bool MemoryMappedGPUBuffer::isLinearBuffer() const
+{
+    return !kUseDRMModifiers || (m_planeCount == 1 && (m_modifier == DRM_FORMAT_MOD_INVALID || m_modifier == DRM_FORMAT_MOD_LINEAR));
+}
+
+bool MemoryMappedGPUBuffer::createDMABufFromGBMBufferObject()
+{
+    ASSERT(m_eglAttributes.isEmpty());
+
+    Vector<UnixFileDescriptor> fds;
+    Vector<uint32_t> offsets;
+    Vector<uint32_t> strides;
+
+    m_eglAttributes = {
+        EGL_WIDTH, static_cast<EGLAttrib>(gbm_bo_get_width(m_bo)),
+        EGL_HEIGHT, static_cast<EGLAttrib>(gbm_bo_get_height(m_bo)),
+        EGL_LINUX_DRM_FOURCC_EXT, static_cast<EGLAttrib>(m_format)
+    };
+
+#define ADD_PLANE_ATTRIBUTES(planeIndex) { \
```

Comment:
Could we achieve the same without a macro? Is there another way to produce the `EGL_DMA_BUF_PLANE##planeIndex##....`  keys? I find this a bit intimidating. 

It wouldn't be a straight replacement as this macro can `return false` and if you turn it into a lambda, its output would need handling...

Reply:
This is a 1:1 copy from another place that defines the same ugly macro - I have a follow-up cleanup PR in the pipeline, that will unify all call sites using this construct, without using macros. Ok to postpone this?

Reply:
yes, thanks! It makes more sense to handle in a follow-up